# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit-test/tests/gc/bug-1698543.js
# Commit: f9ac96a788b0
# Full Hash: f9ac96a788b01dca26ece865efed3a5ee5fd3246
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2021-03-19 09:53:39
# Regressor Bug: 1698543
# File Overlap Count: 7
# Description:
#   Bug 1699364 - Re-add special case pre-barrier for rope flattening r=sfink
#   
#   It turns out that patch for bug 1698543 was too optimistic and we need the special case barrier for rope flattening for another reason: during flattening rope nodes are transformed before their ancestors.  Interior rope nodes are transformed to dependent strings with the base being the root node, and the root transformed into a linear string. Since the root is changed last the GC graph is not safe to traverse until flattening has finished.
#   
#   This makes the test case pass.  I added this test case and the one for the previous bug.  I don't think we need to hide these since this change has only been on nightly so far.
# ==============================================================================

diff -r f2ed6497335f -r f9ac96a788b0 js/src/jit-test/tests/gc/bug-1698543.js
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/js/src/jit-test/tests/gc/bug-1698543.js	Thu Mar 18 19:31:57 2021 +0000
@@ -0,0 +1,32 @@
+// |jit-test| allow-overrecursed; skip-if: !getJitCompilerOptions()['blinterp.enable']
+
+foo = "";
+
+doit(`
+  // XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+  // XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+  function u() { broken(
+  // XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+  // XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+  // XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+  // XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+  // XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+  // XXXXXXXXXXXXXXXXXXXXXXXXXXX
+`);
+
+gczeal(4);
+
+doit("");
+
+unescape(foo);
+
+function doit(x) {
+  try {
+    evaluate(x);
+  } catch (e) {
+    if (e instanceof SyntaxError)
+      doit(x);
+  }
+  x = x.replace(/!/g, "");
+  foo += x + " ";
+}